### PR TITLE
feat: add no-important rule

### PR DESCRIPTION
**Pull request by yannbertrand**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Add a rule to avoid `!important` annotation usage.

#### What changes did you make? (Give an overview)

- Created the rule
- Added it to the recommended rules
- Wrote the docs

#### Related Issues

See #20

#### Is there anything you'd like reviewers to focus on?

Implementation was not complicated but I'm not sure on how much details the docs should include.

(It's my first PR on ESlint, lmk if anything is wrong or if the rule is useless)

### CODE REVIEW

File: src/rules/no-important.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//-----------------------------------------------------------------------------
+// Rule Definition
+//-----------------------------------------------------------------------------
+
+export default {
+	meta: {
+		type: /** @type {const} */ ("problem"),
+
+		docs: {
+			description: "Disallow important annotations",
```
Comment (nzakas) [inline src/rules/no-important.js:15]:
```suggestion
			description: "Disallow !important annotations",
```

File: src/rules/no-important.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+			recommended: true,
+		},
+
+		messages: {
+			forbiddenImportantUsage: "!important annotations are not allowed.",
```
Comment (nzakas) [inline src/rules/no-important.js:20]: Can we shorten this to `unexpectedImportant`?

And we generally like to be a bit gentler with the messages:

```suggestion
			unexpectedImportant: "Unexpected !important annotation found.",
```

File: src/rules/no-important.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+	create(context) {
+		return {
+			"Declaration[important=true]"(node) {
+				context.report({
+					loc: {
+						start: node.loc.start,
+						end: {
+							line: node.loc.start.line,
+							column:
```
Comment (nzakas) [inline src/rules/no-important.js:32]: This will just highlight the property name. I think it would be better to find the location of the `!important` and set that as the location of the error.

File: tests/rules/no-important.test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	},
+	language: "css/css",
+});
+
+ruleTester.run("no-important", rule, {
+	valid: [
+		"a { color: red; }",
+		"a { color: red; background-color: blue; }",
+		"a { color: red; transition: none; }",
+		"body { --custom-property: red; }",
```
Comment (nzakas) [inline tests/rules/no-important.test.js:30]: I'd also add a test with just the word "important":

```suggestion
		"body { --custom-property: red; }",
		"body { --custom-property: important; }",		
```

File: tests/rules/no-important.test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		{
+			code: "a { color: red !important; }",
+			errors: [
+				{
+					messageId: "forbiddenImportantUsage",
+					data: {
+						property: "color",
+						value: "red !important",
+						expected: "red",
+					},
```
Comment (nzakas) [inline tests/rules/no-important.test.js:46]: `data` doesn't apply because you're not passing any `data` to `context.report()` in the rule. Please remove all `data` entries from this file.

File: tests/rules/no-important.test.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+					line: 1,
+					column: 11,
+					endLine: 1,
+					endColumn: 18,
+				},
+			],
+		},
+	],
+});
```
Comment (nzakas) [inline tests/rules/no-important.test.js:89]: Can you also add a test that has multiple properties with `!important` to ensure we're catching them all?

**Review by nzakas — CHANGES_REQUESTED** [review]
Thanks for getting this started, it looks really good. Just a few things to clean up.

File: docs/rules/no-important.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# no-important
+
+Disallow `!important` annotations.
+
+## Background
+
+Needing !important indicates there may be a larger underlying issue.
```
Comment (nzakas) [inline docs/rules/no-important.md:7]: We need more of a description here. :smile: Please see other rule docs for examples.

File: docs/rules/no-important.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+```css
+a {
+	color: red !important;
+}
+
+a .link {
+	font-size: padding: 10px 20px 30px 40px !important;
+}
+```
```
Comment (nzakas) [inline docs/rules/no-important.md:21]: Please also add "Prior Art" section as in other rules.

https://stylelint.io/user-guide/rules/declaration-no-important/
https://github.com/CSSLint/csslint/wiki/Disallow-%21important

File: src/rules/no-important.js (hunk 1 above)
Comment (yannbertrand) [reply src/rules/no-important.js]: Done

File: src/rules/no-important.js (hunk 2 above)
Comment (yannbertrand) [reply src/rules/no-important.js]: Done. Thanks for the suggestions

File: tests/rules/no-important.test.js (hunk 5 above)
Comment (yannbertrand) [reply tests/rules/no-important.test.js]: Missed that one 😅 thanks, it's fixed

**Comment by yannbertrand** [comment]
> Thanks for your work on this. There's some ongoing discussion about whether or not we'd like to have this rule on #20, so we may need to think things through a bit before proceeding.

Sure thing, no problem! Should keep this PR opened? I'll wait for a decision to either get back to work on this one or replace it by another PR depending on the expected scope.

**Comment by nzakas** [comment]
Let's keep this one open for now. I'll just mark it as a draft while we try to figure things out.

**Comment by nzakas** [comment]
@yannbertrand sorry for the delay on this. We'd like to move forward. If you can please take a look at resolving the conflicts and addressing the feedback that was left, we'd appreciate it.

**Comment by nzakas** [comment]
@thecalamiity yes, absolutely. That would be appreciated. :pray:
